### PR TITLE
WIP: Add scrypt as EVP_PKEY_METHOD

### PR DESCRIPTION
**Pull request by johndoe31415**
This adds the scrypt KDF as a EVP_PKEY_METHOD. It is onging work in progress; converted to a PR as of the request by @snhenson.

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by johndoe31415** [comment]
@snhenson Regarding your comment [about the uint64_t parsing](https://github.com/johndoe31415/openssl/commit/ed24fe0e25a9b7c56e50c9174976609fbfef01c4#commitcomment-23313858), is maybe sscanf with the SCNu64 macro defined in <inttypes.h> something we can use? Only came to my mind after-the-fact.

**Comment by snhenson** [comment]
I don't know if there is a portable way to do things. Some of the obscure platforms of interest might not even have sscanf hence my question.

File: include/openssl/kdf.h [hunk 1]
```diff
@@ -28,7 +28,7 @@ extern "C" {
 # define EVP_PKEY_CTRL_SCRYPT_N                 (EVP_PKEY_ALG_CTRL + 10)
 # define EVP_PKEY_CTRL_SCRYPT_R                 (EVP_PKEY_ALG_CTRL + 11)
 # define EVP_PKEY_CTRL_SCRYPT_P                 (EVP_PKEY_ALG_CTRL + 12)
-# define EVP_PKEY_CTRL_SCRYPT_MAXMEM_MIB        (EVP_PKEY_ALG_CTRL + 13)
+# define EVP_PKEY_CTRL_SCRYPT_MAXMEM_BYTES      (EVP_PKEY_ALG_CTRL + 13)
```
Comment (snhenson) [inline include/openssl/kdf.h]: I'd suggest macros for these too. So something like EVP_PKEY_CTX_set_scrypt_N etc. as we do with the TLS PRF. The uint64_t cases maybe a new wrapper function that takes a uint64_t parameter and passes it to the underlying ctrl as a pointer?

**Comment by richsalz** [comment]
Just do the "obvious" ```*10+*p-'0'```  It's what was done for parse_uint64 in evp_test.c

**Comment by johndoe31415** [comment]
@richsalz The parser in evp_test.c isn't entirely correct, however. It will successfully parse 18446744073709551616 as 0 up including 18446744073709551619 (which it parses as 3). For the test I figured this to be okay, but for code that's actively used I thought this might be unwanted. Therefore the own parser.

File: crypto/objects/objects.txt [hunk 2]
```diff
@@ -1524,7 +1524,7 @@ secg-scheme 14 3 : dhSinglePass-cofactorDH-sha512kdf-scheme
 1 3 6 1 4 1 311 60 2 1 3	: jurisdictionC		: jurisdictionCountryName
 
 # SCRYPT algorithm
-1 3 6 1 4 1 11591 4 11		: id-scrypt
+1 3 6 1 4 1 11591 4 11		: scrypt            : id-scrypt
```
Comment (snhenson) [inline crypto/objects/objects.txt]: Typo in comment which I thought I'd fixed. I meant add ": scrypt" at the end. That should keep SN_id_scrypt unchanged (which we should do for compatibility reasons) and add a new LN_id_scrypt.

Comment (johndoe31415) [inline crypto/objects/objects.txt]: Hmmm, still not sure about this. Just tried it like this:

```
1 3 6 1 4 1 11591 4 11          : id-scrypt         : scrypt
```

But then the NID symbol name changes from NID_id_scrypt to NID_scrypt and I get

```
crypto/asn1/p5_scrypt.c: In function ‘pkcs5_scrypt_set’:
crypto/asn1/p5_scrypt.c:202:38: error: ‘NID_id_scrypt’ undeclared (first use in this function)
     keyfunc->algorithm = OBJ_nid2obj(NID_id_scrypt);
```

Is this okay, i.e., should I simply change all occurences of NID_id_scrcypt to NID_scrypt?

**Comment by johndoe31415** [comment]
Rebased to master, because I'm unsure if the build failure is because of something I accidently messed up in objects.txt.

File: crypto/objects/objects.txt (hunk 2 above)
Comment (snhenson) [inline crypto/objects/objects.txt]: Hmm... try this:
```
!Cname id-scrypt
1 3 6 1 4 1 11591 4 11      : id-scrypt : scrypt
```

**Comment by snhenson** [comment]
FYI now #4079 has been implemented using "scrypt" for the name in evp_test and pkeyutl should work.

**Comment by johndoe31415** [comment]
Thank you, @snhenson -- I just rebased to be able to use your changes and will try it our right away!

**Comment by johndoe31415** [comment]
Worked perfectly. Thank you again, @snhenson!

With the major roadblocks out the way, I figure adding PBKDF2 is quite straightforward. If there aren't any objections, I'd get to it then?

Maybe one more question about a corner case: The current way that the PKEY_METHOD scrypt implementation handles passphrases and salts, it isn't possible to specify a zero-length password or salt. Therefore the testcase for this (which was present in testing the scrypt PBE) is commented out for KDF-scrypt. It's an esoteric corner case, should I change the code so that it's possible to set zero-length salt/passphrase for the scrypt KDF as well?

File: test/recipes/30-test_evp_data/evpkdf.txt [hunk 3]
```diff
@@ -252,4 +252,63 @@ Ctrl.IKM = hexkey:0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c
...
+
+# Empty password/salt currently unsupported by PKEY KDF scrypt
+#KDF = scrypt
+#Ctrl.pass = pass:""
+#Ctrl.salt = salt:""
+#Ctrl.N = N:16
+#Ctrl.r = r:1
+#Ctrl.p = p:1
+#Output = 77d6576238657b203b19ca42c18a0497f16b4844e3074ae8dfdffa3fede21442fcd0069ded0948f8326a753a0fc81f17e8d3e0fb2e0d3628cf35e20c38d18906
+
```
Comment (snhenson) [inline test/recipes/30-test_evp_data/evpkdf.txt]: I think it's important that we find a way to support this. The reason is that in future I'd like to make EVP_PKEY_METHOD the preferred interface for scrypt (and other PBE algorithms) so it must be at least as flexible as the underlying function. I can think of several ways to do that, one is to allocate the fields when set even if they're zero length so the != NULL check passes.

**Comment by snhenson** [comment]
This is looking closer to something which can be included. Couple of additions needed though: this needs to be documented. Also some of the commits can be squashed (and log messages edited) and the whole thing rebased against the current master.

**Comment by snhenson** [comment]
I think PBKDF2 should be easier to handle: in some ways scrypt is the most difficult case.

File: test/recipes/30-test_evp_data/evpkdf.txt (hunk 3 above)
Comment (johndoe31415) [inline test/recipes/30-test_evp_data/evpkdf.txt]: Done for PBKDF2 right off the bat and for scrypt now as well with 305df7b.

**Comment by johndoe31415** [comment]
Hi @snhenson, that's great news! I've already rebased to master, will write some documentation as well and then squash the whole scrypt odysee in one commit, the PBKDF2 as another.

The thing I'm not feeling entirely comfortable with right now is the macro setters, which don't have any tests. I'd like to write a small example program that at least tests one vector for each manually. This would ensure that stupid mistakes (copy/paste kind of stuff or some erroneous typo in one of the header files) gets caught right away. I'm writing this code anyways for the "EXAMPLE" section of the manpage, so I figure we could as well include it as a test if that's okay?

Thanks again so much for your support!

File: include/openssl/kdf.h [hunk 4]
```diff
@@ -59,6 +68,48 @@ extern "C" {
...
                               EVP_PKEY_CTRL_HKDF_MODE, mode, NULL)
+
+# define EVP_PKEY_CTX_set1_scrypt_pass(pctx, pass, passlen) \
+            EVP_PKEY_CTX_ctrl(pctx, -1, EVP_PKEY_OP_DERIVE, \
+                            EVP_PKEY_CTRL_PASS, passlen, (void *)(pass))
+
+# define EVP_PKEY_CTX_set1_scrypt_salt(pctx, salt, saltlen) \
+            EVP_PKEY_CTX_ctrl(pctx, -1, EVP_PKEY_OP_DERIVE, \
+                            EVP_PKEY_CTRL_SCRYPT_SALT, saltlen, (void *)(salt))
+
```
Comment (snhenson) [inline include/openssl/kdf.h]: Suggest we have a common name for this without reference to the algorithm. For example EVP_PKEY_CTX_set1_pbe_pass() the expectation here is that *all* PBE KDFs will support it.

File: test/recipes/30-test_evp_data/evpkdf.txt [hunk 5]
```diff
@@ -252,4 +252,213 @@ Ctrl.IKM = hexkey:0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c0c
...
+KDF = scrypt
+Ctrl.pass = pass:pleaseletmein
+Ctrl.salt = salt:SodiumChloride
+Ctrl.N = N:1048576
+Ctrl.r = r:8
+Ctrl.p = p:1
+Ctrl.maxmem_bytes = maxmem_bytes:10000000000
+Output = 2101cb9b6a511aaeaddbbe09cf70f881ec568d574a2ffd4dabe5ee9820adaa478e56fd8f4ba5d09ffa1c6d927c40f4c337304049e8a952fbcbf45c6fa77a41a4
+
```
Comment (snhenson) [inline test/recipes/30-test_evp_data/evpkdf.txt]: I don't think it's a good idea to use the above by default. It requires 1GB+ of memory and on some systems that could grind them to a halt or cause "make test" to fail with an out of memory error. It did that on my VM when I was first testing scrypt.

**Comment by snhenson** [comment]
Can you exclude the PBKDF2 KDF from this PR? That can be a separate PR once this is merged.

File: include/openssl/kdf.h (hunk 4 above)
Comment (johndoe31415) [inline include/openssl/kdf.h]: Fixed in be01d24384. In your comment you wrote about pbe_pass(), but the highlighted code shows scrypt_salt(). Do you suggest that I rename the _salt() functions as well? For now, the patch only introduces pbe_pass().

File: test/recipes/30-test_evp_data/evpkdf.txt (hunk 5 above)
Comment (johndoe31415) [inline test/recipes/30-test_evp_data/evpkdf.txt]: Fixed in ee41de12af.

**Comment by johndoe31415** [comment]
Rebased to master and squashed everything into three commits (the original contribution, the documentation and the new test file). Removed PBKDF2 entirely from this PR and extracted it into https://github.com/openssl/openssl/pull/4087.

File: crypto/kdf/scrypt.c [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+    size_t salt_len;
+    uint64_t N, r, p;
+    uint64_t maxmem_bytes;
+} SCRYPT_PKEY_CTX;
+
+/* Custom uint64_t parser since we do not have strtoull */
+static int atou64(const char *nptr, uint64_t *result)
+{
+    uint64_t value = 0;
+    while (*nptr) {
```
Comment (snhenson) [inline crypto/kdf/scrypt.c]: Coding style says blank line after declarations.

File: crypto/kdf/scrypt.c [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+    kctx->maxmem_bytes = 1025 * 1024 * 1024;
+
+    ctx->data = kctx;
+
+    return 1;
+}
+
+static void pkey_scrypt_cleanup(EVP_PKEY_CTX *ctx)
+{
+    SCRYPT_PKEY_CTX *kctx = ctx->data;
```
Comment (snhenson) [inline crypto/kdf/scrypt.c]: Add blank line.

File: crypto/kdf/scrypt.c [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+        return -2;
+
+    }
+}
+
+static int pkey_scrypt_ctrl_uint64(EVP_PKEY_CTX *ctx, int type,
+                                   const char *value)
+{
+    uint64_t int_value;
+    if (!atou64(value, &int_value)) {
```
Comment (snhenson) [inline crypto/kdf/scrypt.c]: Ditto

File: include/openssl/kdf.h (hunk 4 above)
Comment (snhenson) [inline include/openssl/kdf.h]: No I think the salt can stay as it is. The salt (or at least how it is encoded) can vary from one PBE algorithm to another.

File: doc/man3/EVP_PKEY_CTX_set1_scrypt_salt.pod [hunk 9]
```diff
@@ -0,0 +1,158 @@
+=pod
```
Comment (snhenson) [inline doc/man3/EVP_PKEY_CTX_set1_scrypt_salt.pod]: I'd suggest the documentation follows the style of other algorithms like RSA-PSS: that is located in man7 and named after the algorithm itself. There are some existing KDFs (like the TLS1 PRF) that don't yet follow this but they will be updated at some point.

**Comment by johndoe31415** [comment]
Hi @snhenson, thanks for your review. I've addressed all of your comments in c0deff0.

File: crypto/objects/obj_mac.num [hunk 10]
```diff
@@ -1117,3 +1117,4 @@ RSA_SHA3_224		1116
 RSA_SHA3_256		1117
 RSA_SHA3_284		1118
 RSA_SHA3_512		1119
+scrypt		1120
```
Comment (snhenson) [inline crypto/objects/obj_mac.num]: Is this left over from a previous version? On my system after making the changes to objects.txt and doing `make update` I get no changes in obj_mac.num

Comment (johndoe31415) [inline crypto/objects/obj_mac.num]: Well spotted. I've reverted

```
    crypto/asn1/charmap.h \
    crypto/conf/conf_def.h \
    crypto/objects/obj_dat.h \
    crypto/objects/obj_mac.num \
    crypto/objects/obj_xref.h \
    crypto/objects/obj_xref.txt \
    include/openssl/obj_mac.h
```

and ran ``make update``. Can confirm that this is a leftover from the previous runs that became obsolete. b7c9b69 fixes this.

**Comment by snhenson** [comment]
Approved subject to squashing the last two fixes. Many thanks for your contribution!

**Comment by johndoe31415** [comment]
Hi @snhenson, that's awesome news! I've squashed/fixuped the last hiccups. Thanks again for your patience. Cheers, Johannes

**Comment by snhenson** [comment]
Pushed now, thanks everyone.

File: crypto/evp/pmeth_lib.c [hunk 11]
```diff
@@ -355,6 +359,12 @@ int EVP_PKEY_CTX_ctrl(EVP_PKEY_CTX *ctx, int keytype, int optype,
 
 }
 
+int EVP_PKEY_CTX_ctrl_uint64(EVP_PKEY_CTX *ctx, int keytype, int optype,
+                                int cmd, uint64_t value)
```
Comment (dot-asm) [inline crypto/evp/pmeth_lib.c]: Formatting nit. There is corresponding misalignment in include/openssl/evp.h. It wouldn't be inappropriate to fix it in new request, @johndoe31415.

Comment (johndoe31415) [inline crypto/evp/pmeth_lib.c]: Ah, dang! As requested, new PR: https://github.com/openssl/openssl/pull/4166 -- thanks for the review!

**Pull request closed without merging**